index.bs

```diff
@@ -383,6 +383,41 @@ This API is disabled by default in all cross-origin frames using the [[#permissi
+
+Issue: Document operations susceptible to out-of-bounds access as a guidance to implementers.
+
+As a future-proofing, the API design allows certain operation (annotated with "can be generically emulated") defined in this specification to be deprecated for security, performance, or other reasons without breaking compability. This is made possible by high-level functions that are defined in terms of smaller primitive operations defined in this specifications. This enables a native implementation of a high-level function to be replaced with a polyfill implementation.
+
+Issue: Investigate side channel attack feasibility considering the current state where CPU is shared between processes running renderers.
+
+In order to not allow an attacker to target a specific implementation that may contain a flaw, the [[#programming-model-device-selection]] mechanism is a hint only, and the concrete device selection is left to the implementation. As a further mitigation, no device enumeration mechanism is defined.
+
+Issue: Hinting partially mitigates the concern. Investigate additional mitigations.
+
+The API design minimizes the attack surface for the compiled computational graph. The {{MLGraphBuilder}} interface that hosts the various operations is a data definition API and as such doesn't execute anything, only constructs data. What follows, is that the potential for an attack is limited to when binding the data to the graph before executing it by invoking the {{MLGraph/compute()}} method. This enables implementers to focus on hardening the {{MLGraph/compute()}} method. For example, by making sure it honors the boundary of data and fails appropriately when the bounds are not respected.
+
+Purpose-built Web APIs for measuring high-resolution time mitigate against timing attacks using techniques such as resolution reduction, adding jitter, detection of abuse and API call throttling [[hr-time-3]]. It is expected a typical multi-process architecture implementation of the WebNN API makes use of inter-process communication (IPC) mechanism that inherently adds additional jitter on top of the built-in high-resolution time mitigations. Regardless of these mitigations, implementers are always advised to consider and test their implementations against timing attacks, and in particular if IPC is not used.
+
+## Guidelines for new operations ## {#security-new-ops}
+
+To ensure operations defined in this specification are shaped in a way they can be implemented securely, this section includes guidelines on how operations are expected to be defined to reduce potential for implementation problems. These guidelines are expected to evolve over time to align with industry best practices:
+
+- Avoid reshaping tensors
```

Comment:
I am not quite sure about this one. Do we have any references, say ONNX document?

Reply:
```suggestion
```

The [ONNX document](https://lists.w3.org/Archives/Public/www-archive/2021Nov/att-0000/W3C_Adding_new_Operators.pdf) does not mention this. I believe this must be my misinterpretation/oversimplification of [our discussion](https://www.w3.org/2022/01/27-webmachinelearning-minutes.html#t03).

I'm suggesting we drop this bullet unless we can reword this to make sense in this context.

Reply:
Reshaping is quite common in an ML graph. I'm not entirely sure why it should be discouraged from a security standpoint.

Reply:
It sounds like we'll simply drop this bullet as suggested.